lib/irb/type_completion/types.rb

```diff
@@ -0,0 +1,443 @@
+          method = rbs_search_method klass, method_name, singleton
+          next [] unless method
+          method.method_types.map do |method_type|
+            score = 0
+            score += 2 if !!method_type.block == has_block
+            reqs = method_type.type.required_positionals
+            opts = method_type.type.optional_positionals
+            rest = method_type.type.rest_positionals
+            trailings = method_type.type.trailing_positionals
+            keyreqs = method_type.type.required_keywords
+            keyopts = method_type.type.optional_keywords
+            keyrest = method_type.type.rest_keywords
+            args = args_types
+            if kwargs_type&.any? && keyreqs.empty? && keyopts.empty? && keyrest.nil?
+              kw_value_type = UnionType[*kwargs_type.values]
+              args += [InstanceType.new(Hash, K: SYMBOL, V: kw_value_type)]
+            end
+            if has_splat
+              score += 1 if args.count { !(_1.is_a? Splat) } <= reqs.size + opts.size + trailings.size
+            elsif reqs.size + trailings.size <= args.size && (rest || args.size <= reqs.size + opts.size + trailings.size)
+              score += 2
+              centers = args[reqs.size...-trailings.size]
+              given = args.first(reqs.size) + centers.take(opts.size) + args.last(trailings.size)
+              expected = (reqs + opts.take(centers.size) + trailings).map(&:type)
+              if rest
+                given << UnionType[*centers.drop(opts.size)]
+                expected << rest.type
+              end
+              if given.any?
+                score += given.zip(expected).count do |t, e|
+                  e = from_rbs_type e, receiver_type
+                  intersect?(t, e) || (intersect?(STRING, e) && t.methods.include?(:to_str)) || (intersect?(INTEGER, e) && t.methods.include?(:to_int)) || (intersect?(ARRAY, e) && t.methods.include?(:to_ary))
+                end.fdiv(given.size)
+              end
+            end
+            [[method_type, given || [], expected || []], score]
+          end
+        end
+        max_score = methods_with_score.map(&:last).max
+        methods_with_score.select { _2 == max_score }.map(&:first)
+      end
+
+      def self.intersect?(a, b)
+        atypes = a.types.group_by(&:class)
+        btypes = b.types.group_by(&:class)
+        if atypes[SingletonType] && btypes[SingletonType]
+          aa, bb = [atypes, btypes].map {|types| types[SingletonType].map(&:module_or_class) }
+          return true if (aa & bb).any?
+        end
+
+        aa, bb = [atypes, btypes].map {|types| (types[InstanceType] || []).map(&:klass) }
+        (aa.flat_map(&:ancestors) & bb).any?
+      end
+
+      def self.type_from_object(object)
+        case object
+        when Array, Hash, Module
+          type_from_object_recursive(object, max_level: 4)
+        else
+          klass = object.singleton_class rescue object.class
```

Comment:
If the object inherits directly from `BasicObject`, this will cause error because it likely won't have neither `singleton_class` nor `class` method. Please add a few cases for `BasicObject` because they actually appear more often than we thought.
Also, perhaps it's better to implement the fallback mechanism at this level?

Reply:
I completely forgot about BasicObject. fixed in 63693d3268f3ef1fa559badc0e2f2159b181cb10

Fallback mechanism like `BaseScope.type_of(fallback: Types::NIL){b.local_variable_get name}` is actually a fallback of `local_variable_get(name)` causing NameError.

I think type_from_object normally does not raise error after BasicObject fix.
I changed to explicitly rescue the expected error and remove fallback mechanism, because current fallback will suppress and might hide important error.
```
BaseScope.type_of(fallback: fallback) { @binding.local_variable_get(name) }
# ↓
value = begin
  @binding.local_variable_get(name)
rescue NameError
end
Types.type_from_object(value)
```
8a0fb2e4743a9b97590272123ee2c6289393383f